src/engine/AbsoluteValueConstraint.h

```diff
@@ -0,0 +1,171 @@
+
+class AbsoluteValueConstraint : public PiecewiseLinearConstraint
+{
+
+public:
+    enum PhaseStatus {
+        PHASE_NOT_FIXED = 0,
+        PHASE_POSITIVE = 1,
+        PHASE_NEGATIVE = 2,
+    };
+
+    AbsoluteValueConstraint( unsigned b, unsigned f );
+
+    /*
+      Return a clone of the constraint.
+    */
+    PiecewiseLinearConstraint *duplicateConstraint() const;
+
+    /*
+      Restore the state of this constraint from the given one.
+    */
+    void restoreState( const PiecewiseLinearConstraint *state );
+
+    /*
+      Register/unregister the constraint with a talbeau.
+     */
+    void registerAsWatcher( ITableau *tableau);
+    void unregisterAsWatcher( ITableau *tableau);
+
+    /*
+      These callbacks are invoked when a watched variable's value
+      changes, or when its bounds change.
+    */
+    void notifyVariableValue( unsigned variable, double value );
+    void notifyLowerBound( unsigned variable, double bound );
+    void notifyUpperBound( unsigned variable, double bound );
+
+    /*
+       Returns true iff the variable participates in this piecewise
+       linear constraint.
+     */
+    bool participatingVariable( unsigned variable ) const;
+
+    /*
+      Get the list of variables participating in this constraint.
+    */
+    List<unsigned> getParticipatingVariables() const;
+
+    /*
+      Returns true iff the assignment satisfies the constraint
+    */
+    bool satisfied() const;
+
+    /*
+      Returns a list of possible fixes for the violated constraint.
+    */
+    List<PiecewiseLinearConstraint::Fix> getPossibleFixes() const;
+
+    /*
+      Return a list of smart fixes for violated constraint.
+      Currently not implemented, just calls getPossibleFixes().
+    */
+    List<PiecewiseLinearConstraint::Fix> getSmartFixes( ITableau *tableau ) const;
+
+    /*
+      Returns the list of case splits that this piecewise linear
+      constraint breaks into. These splits need to complementary,
+      i.e. if the list is {l1, l2, ..., ln-1, ln},
+      then ~l1 /\ ~l2 /\ ... /\ ~ln-1 --> ln.
+    */
+    List<PiecewiseLinearCaseSplit> getCaseSplits() const;
+
+    /*
+      Check whether the constraint's phase has been fixed.
+    */
+    void fixPhaseIfNeeded();
+    bool phaseFixed() const;
+
+    /*
+     * If the constraint's phase has been fixed, get the (valid) case split.
+     */
+    PiecewiseLinearCaseSplit getValidCaseSplit() const;
+
+    /*
+      Preprocessing related functions, to inform that a variable has
+      been eliminated completely because it was fixed to some value,
+      or that a variable's index has changed (e.g., x4 is now called
+      x2). constraintObsolete() returns true iff and the constraint
+      has become obsolote as a result of variable eliminations.
+     */
+    void eliminateVariable( unsigned variable, double fixedValue );
+    void updateVariableIndex( unsigned oldIndex, unsigned newIndex );
+    bool constraintObsolete() const;
```

Comment:
Maybe name the bool returning test functions with is prefix, i.e. isConstraintObsolete()? It would distinguish from imperative function names and make the code easier to read.

Reply:
Currently this convention is not really followed, not even in this particular class. Is it in our coding guidelines?